### ISSUE TITLE
Probably no dropout applied to the encoder in the RNN encoder/decoder ASR

### ISSUE THREAD

**Issue by qmeeus** [espnet/espnet#1780]
Hi,

When trying to apply dropout on the rnn encoder, I get this warning:
```
/lib/python3.6/site-packages/torch/nn/modules/rnn.py:50: UserWarning: dropout option adds dropout after all but last recurrent layer, so non-zero dropout expects num_layers greater than 1, but got dropout=0.5 and num_layers=1
  "num_layers={}".format(dropout, num_layers))
```
Indeed, [after a quick check in the code](https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/rnn/encoders.py#L35):
```python
rnn = torch.nn.LSTM(inputdim, cdim, dropout=dropout, num_layers=1, bidirectional=bidir,
                                batch_first=True) if "lstm" in typ \
                else torch.nn.GRU(inputdim, cdim, dropout=dropout, num_layers=1, bidirectional=bidir, batch_first=True)
```

**Comment by b-flo** [comment]
Hi,

Thanks for the report. We are aware of the problem: dropout won't be applied if `use_projection=True`, as RNNP is a stack of 1-layer torch.nn.LSTM and dropout is disabled if `num_layers=1`.

@sw005320 It was already reported in multiple discussions but I'm not sure why we didn't fix that? There may have been a reason.

**Comment by qmeeus** [comment]
I wrote a fix, I will do a merge request after testing

**Comment by sw005320** [comment]
> @sw005320 It was already reported in multiple discussions but I'm not sure why we didn't fix that? There may have been a reason.

It should be fixed...

> I wrote a fix, I will do a merge request after testing

That is really helpful.
It would be great if you make a PR about it.

**Comment by qmeeus** [comment]
Here you go
https://github.com/espnet/espnet/pull/1784

**Comment by qmeeus** [comment]
Before closing this, is it intentional to use the same dropout rate for the attention layer and the encoder? If not, I suggest adding a dropout rate specific to the attention layer and renaming the argument "dropout_rate" to "encoder_dropout" or something like that

**Comment by sw005320** [comment]
I think this is a good idea in general. Personally, I did not have good success in introducing dropout anyway and making dropout related development items in a lower priority due to that. But I think this function is quite reasonable.

**Comment by qmeeus** [comment]
My model overfits on the encoder side (CTC) but if I'm using dropout in the attention, the decoder is not able to attend to the right parts of the input anymore. I will look into it
